### PR TITLE
[IMP] utm: ignore capital letter

### PR DESCRIPTION
**Pull request by datonefaridze**
To avoid creating duplicates repeatedly sources and mediums won't be
created automatically if there's a case-insensitive match in the
existing records.

We also update the name for case-incensitive duplicates for records
created in the back end so users should notice if they create a
case-incensitive duplicate the same way they did if they created an
exact duplicate

Some tests were also modified to test the updated functionality.

task-3329888

Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/utm/models/utm_mixin.py [hunk 1]
```diff
@@ -73,6 +73,12 @@ def _find_or_create_record(self, model_name, name):
 
         return record
 
+    def _find_or_create_lower_record(self, model_name, vals_list, seach_domain):
+        existing_ids = {vals['id']: vals['name'].lower() for vals in self.env[model_name].search_read(seach_domain, ['name'])}
+        existing_records = self.env[model_name].browse(existing_ids.keys())
+        new_vals_list = [x for x in vals_list if x['name'].lower() not in existing_ids.values()]
```
Comment (reth-odoo) [inline addons/utm/models/utm_mixin.py:79]:
```suggestion
        existing_names_lower = existing_records.mapped(lambda record: record.name.lower())
        new_vals_list = [vals for vals in vals_list if vals['name'].lower() in existing_names_lower]
```

File: addons/utm/models/utm_source.py [hunk 2]
```diff
@@ -17,10 +18,14 @@ class UtmSource(models.Model):
 
     @api.model_create_multi
     def create(self, vals_list):
-        new_names = self.env['utm.mixin']._get_unique_names(self._name, [vals.get('name') for vals in vals_list])
+        seach_domain = expression.OR([[('name', '=ilike', name['name'].lower())] for name in vals_list])
+        existing_records, new_vals_list = self.env['utm.mixin']._find_or_create_lower_record('utm.source', vals_list, seach_domain)
```
Comment (reth-odoo) [inline addons/utm/models/utm_source.py:22]: Same here then, if we're doing the same expression.OR twice and are using a new method, might as well do that in the method itself.

File: addons/utm/models/utm_source.py [hunk 3]
```diff
@@ -17,10 +18,14 @@ class UtmSource(models.Model):
...
+        new_names = self.env['utm.mixin']._get_unique_names(self._name, [vals.get('name') for vals in new_vals_list])
         for vals, new_name in zip(vals_list, new_names):
             vals['name'] = new_name
-        return super().create(vals_list)
+        new_records = super().create(new_vals_list)
+        total_records = new_records + existing_records
+        return total_records
```
Comment (reth-odoo) [inline addons/utm/models/utm_source.py:28]: Same as above

**Review by reth-odoo — COMMENTED** [review]
Some general form for the early review request.

Seems like tde is suggesting to just do this when creating utm values dynamically from the mixin. That sounds sensible to me, maybe wait for awa to respond to that to see how to proceed and recheck with lna if just that would be alright.

Probably much less annoying than having the create method dynamically modify its values, even though effectively we already do that to avoid duplicates.

But doing it here should be ok too.

**Comment by reth-odoo** [comment]
@datonefaridze please refer to the chatter on the task internally :)

They've given reasons not to bother with data manipulation

File: addons/utm/models/utm_medium.py [hunk 4]
```diff
@@ -20,10 +21,14 @@ class UtmMedium(models.Model):
 
     @api.model_create_multi
     def create(self, vals_list):
-        new_names = self.env['utm.mixin']._get_unique_names(self._name, [vals.get('name') for vals in vals_list])
-        for vals, new_name in zip(vals_list, new_names):
+        seach_domain = expression.OR([[('name', '=ilike', name['name'].lower())] for name in vals_list])
+        existing_records, new_vals_list = self.env['utm.mixin']._find_or_create_lower_record('utm.medium', vals_list, seach_domain)
```
[the remark this answers is not in the document]
Comment (datonefaridze) [reply addons/utm/models/utm_medium.py:25]: > augment it inside _find_or_create_lower_record
why would i need to augment that domain? domain doesn't need to change at all, you just pass domain and that's it.

and i don't think i can use `_find_or_create_record` because it's very different from my goal, it doesn't work on case insensitive manner.

@reth-odoo

File: addons/utm/models/utm_source.py (hunk 2 above)
Comment (datonefaridze) [reply addons/utm/models/utm_source.py:22]: @reth-odoo what i meant was: what if i am have different domain (instead of =ilike?) i think having domain outside gives much more flexibility

**Comment by datonefaridze** [comment]
@reth-odoo don't u think i have to delete this line: `new_names = self.env['utm.mixin']._get_unique_names(self._name, [vals.get('name') for vals in new_vals_list])` ? This line in responsible for increasing utm count. Email, and if you create Email again it will create Email [2]. basically to avoid duplication, but my script already attaches newly created record to already existing ones, so i don't think duplication will ever happen. (maybe we can get defensive and still leave it alone, but idk)

**Comment by datonefaridze** [comment]
i know it has test problems, i am just waiting for indian review team so that i can correct tests (or meanwhile i might also do it)

File: addons/utm/tests/test_utm.py [hunk 5]
```diff
@@ -34,10 +34,11 @@ def test_name_generation(self):
```
Comment (reth-odoo) [inline addons/utm/tests/test_utm.py:34]: Doesn't need to be its own commit

**Review by reth-odoo — COMMENTED** [review]
@datonefaridze 
It feels a bit complicated for what we're trying to achieve.

I proposed a lighter solution in the comments, let me know what you think :)

File: addons/utm/tests/test_utm.py [hunk 6]
```diff
@@ -56,19 +57,50 @@ def test_name_generation(self):
             self.assertEqual(utm_1.name, 'UTM 1', msg='This name is already unique')
             self.assertEqual(utm_2.name, 'UTM 2', msg='This name is already unique')
             self.assertEqual(utm_3.name, 'UTM dup [2]', msg='Must add a counter as suffix to ensure uniqueness')
-            self.assertEqual(utm_4.name, 'UTM dup [3]', msg='Must add a counter as suffix to ensure uniqueness')
-            self.assertEqual(utm_5.name, 'UTM dup [4]', msg='Must add a counter as suffix to ensure uniqueness')
```
[the remark this answers is not in the document]
Comment (datonefaridze) [reply addons/utm/tests/test_utm.py:60]: i will see other comments, currently i am working on another task (i already hate this one :) for some reason it's complicated and unnecessary )

File: addons/utm/tests/test_utm.py [hunk 7]
```diff
@@ -34,10 +34,11 @@ def test_name_generation(self):
...
         A counter must be added at the end of the name if it's not the case.
         """
-        for utm_model in ('utm.source', 'utm.medium', 'utm.campaign'):
+        for utm_model in ('utm.source', 'utm.medium'):
             utm_0 = self.env[utm_model].create({'name': 'UTM dup'})
-
-            utm_1, utm_2, utm_3, utm_4, utm_5 = self.env[utm_model].create([{
+            # Because we avoid case insensitive values,
+            # we expect only three utms to be created
+            utm_1, utm_2, utm_3 = self.env[utm_model].create([{
```
[the remark this answers is not in the document]
Comment (datonefaridze) [reply addons/utm/tests/test_utm.py:41]: yeah i know just ignore tests currently, it's a garbage, i left it cuz i was constantly changing a code. i will improve it now

File: addons/utm/tests/test_utm.py (hunk 5 above)
Comment (datonefaridze) [reply addons/utm/tests/test_utm.py:34]: why? i think explaining why tests were modified is good thing. but i will remove if you insist

File: addons/utm/models/utm_medium.py [hunk 8]
```diff
@@ -20,10 +21,12 @@ class UtmMedium(models.Model):
...
     def create(self, vals_list):
-        new_names = self.env['utm.mixin']._get_unique_names(self._name, [vals.get('name') for vals in vals_list])
-        for vals, new_name in zip(vals_list, new_names):
+        seach_domain = expression.OR([[('name', '=ilike', val['name'].lower())] for val in vals_list])
+        existing_records, new_vals_list = self.env['utm.mixin']._find_or_create_lower_record('utm.medium', vals_list, seach_domain)
+        new_names = self.env['utm.mixin']._get_unique_names(self._name, [vals.get('name') for vals in new_vals_list])
+        for vals, new_name in zip(new_vals_list, new_names):
             vals['name'] = new_name
-        return super().create(vals_list)
+        return super().create(new_vals_list) + existing_records
```
[the remark this answers is not in the document]
Comment (datonefaridze) [reply addons/utm/models/utm_medium.py:29]: @reth-odoo why do we have to lower the names i don't get that part ( also couldn't find a place where it lowers the lketters, but assuming it does i don't understand why we should do it, people should be able to create FaceBook right?)
also `_find_or_create_record` doesn't use ilike (perhaps you meant `_get_unique_names` ? )

[the remark this answers is not in the document]
Comment (reth-odoo) [reply addons/utm/models/utm_medium.py:29]: > _find_or_create_record doesn't use ilike

It was a statement for how things would be ^^
So it doesn't now, but it should. If we find any matching we say that's the one, regardless of case.
If none exists, we create it in lowercase. It ensures consistency at not extra cost.

> Should be able to create FaceBook

Then don't lowercase it but still consider it a duplicate and create a new record with `Facebook` `FaceBook [2]`
Or throw a validation error.

It's really weird to have a create return an existing record actually because the front-end, notably, will expect that create is a new record. So if you go to create your new source manually, and it already exists, all your values are changed to the ones of the original record.

File: addons/utm/tests/test_utm.py (hunk 5 above)
Comment (reth-odoo) [reply addons/utm/tests/test_utm.py:34]: Well the commits should make sense independently as much as possible. Here:

if you take only the commit that makes the functional change -> this test breaks
if you take only this test -> it doesn't really make sense to change this

If the reason for the change in tests isn't obvious you can just make the commit message longer in the original commit :)

File: addons/utm/models/utm_medium.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (datonefaridze) [reply addons/utm/models/utm_medium.py:29]: @reth-odoo i was also hesitant on create method returning already existing record, but then in UI it looks strange if i don't do it, see [this](https://www.awesomescreenshot.com/video/24972093?key=b8e3abf1146a15edf734b4ae8a789f03) video

I think you wrote a good idea and i will think about it, i have read it 100 times but my brain is exploding cuz it's overloaded, i will read it tomorrow

[the remark this answers is not in the document]
Comment (datonefaridze) [reply addons/utm/models/utm_medium.py:29]: @reth-odoo 

> Then don't lowercase it but still consider it a duplicate and create a new record with Facebook FaceBook [2]
Or throw a validation error.

i don't think that's right, because i think the whole point of ignoring letter cases is to have better reporting. So Facebook and Facebook [1] are basically the same thing, also Facebook and FACEBOOK

[the remark this answers is not in the document]
Comment (reth-odoo) [reply addons/utm/models/utm_medium.py:29]: If it's automatically created it won't happen, and if it's not people will notice that it's a duplicate thanks to the name.

But if you don't like the idea, I prefer validation error over silently returning an existing record in create

**Comment by datonefaridze** [comment]
@reth-odoo what do u think of current solution? i think we shouldn't change `_get_unique_names`, if people want to manually create same utms let them have it. what do u think?

File: addons/utm/models/utm_mixin.py [hunk 9]
```diff
@@ -62,7 +62,7 @@ def _find_or_create_record(self, model_name, name):
         """Based on the model name and on the name of the record, retrieve the corresponding record or create it."""
         Model = self.env[model_name]
 
-        record = Model.with_context(active_test=False).search([('name', '=', name)], limit=1)
+        record = Model.with_context(active_test=False).search([('name', '=ilike', name)], limit=1)
```
Comment (reth-odoo) [inline addons/utm/models/utm_mixin.py:65]: problem with this is that if I create both `FaceBook` and `Facebook`, it will just pick whatever happens to be first in DB instead of the correct campaign.

Ideally, we wouldn't change this to ilike at all, so older records can still be uniquely identified.
~~But if not, we should remove the limit on the search and order by exact equality, then pick the first. So that if there's one exactly equal, we pick that one, otherwise we pick a similar one.~~ <- that's one fix to this particular issue but I'm not a fan and it doesn't do anything for backend

I think a good way to mitigate this is to ensure we have unique names case-insensitively. i.e. either enforce lowercase or otherwise make `_get_unique_names` add `[2]` even if the records don't have the name name case-sensitively. So either of:
- `facebook`, `facebook [1]`
- `FaceBook`, `Facebook [1]`

Comment (datonefaridze) [reply addons/utm/models/utm_mixin.py:65]: @reth-odoo what do u think of this?

**Review by reth-odoo — COMMENTED** [review]
~~text~~ rediscussing with lna

**Review by reth-odoo — COMMENTED** [review]
As discussed, after rechecking the spec with lna the approach now seems ok.

Just needs a couple tests and we're good :)

**Comment by datonefaridze** [comment]
@reth-odoo hi, what do u think about this?

File: addons/utm/models/utm_mixin.py [hunk 10]
```diff
@@ -85,6 +85,11 @@ def _get_unique_names(self, model_name, names):
             Input: ['test', 'test [3]', 'bob', 'test', 'test']
             Output: ['test [2]', 'test [3]', 'bob', 'test [4]', 'test [5]']
 
+        The function also works case insensitively.
+        E.G.
+            Assume only UTM that exists in the database is "TeST"
+            Input: ['test', 'tesT [3]', 'bob', 'Test', 'test']
+            Output: ['test [2]', 'tesT [3]', 'bob', 'Test [4]', 'test [5]']
```
Comment (reth-odoo) [inline addons/utm/models/utm_mixin.py:92]: Maybe we can just update the examples above to avoid adding a bunch of lines just for this

**Review by reth-odoo — APPROVED** [review]
- Please do it all in one commit
- Commit message can be updated to state the purpose e.g. 
- "to avoid creating duplicates repeatedly sources and mediums won't be created automatically if there's a case-insensitive match in the existing records"
- "we also update the name for case-incensitive duplicates for records created in the back end so users should notice if they create a case-incensitive duplicate the same way they did if they created an exact duplicate"

Once updated, good for me :)
Greening already

File: addons/utm/models/utm_mixin.py (hunk 10 above)
Comment (datonefaridze) [reply addons/utm/models/utm_mixin.py:92]: @reth-odoo i will do in one commit, but that is confusing. because sometimes people tell me to have test and feature in different commits, sometimes in one, i don't get which situation is which

Comment (reth-odoo) [reply addons/utm/models/utm_mixin.py:92]: I've never seen a PR with tests in a different commit. Not saying it can't happen but I don't see a reason for it here at least

Comment (tde-banana-odoo) [reply addons/utm/models/utm_mixin.py:92]: > @reth-odoo i will do in one commit, but that is confusing. because sometimes people tell me to have test and feature in different commits, sometimes in one, i don't get which situation is which

Would say: if you are going to change something that is quite important and it is currently not tested, sometimes it is a good idea to 1/ write tests (and commit), 2/ change a behavior and update tests. Otherwise we generally do both at the same time. Here I would say we don't need to separate commits between feature and tests.

File: addons/website_links/tests/test_ui.py [hunk 11]
```diff
@@ -20,7 +20,7 @@ def _get_title_from_url(addr, **kw):
     def test_01_test_ui(self):
         self.env['link.tracker'].search_or_create({
             'campaign_id': self.env['utm.campaign'].create({'name': 'Sale'}).id,
-            'medium_id': self.env['utm.medium'].create({'name': 'Website'}).id,
+            'medium_id': self.env['utm.medium'].create({'name': 'Test Website'}).id,
             'source_id': self.env['utm.source'].create({'name': 'Search'}).id,
```
Comment (datonefaridze) [inline addons/website_links/tests/test_ui.py:24]: @reth-odoo can you please review it again? especially this file. maybe i can also ask to tde? this file didn't make much sense to me, because: why do we create search, when we will only use search engine.

File: addons/mass_mailing/tests/test_mailing_internals.py [hunk 12]
```diff
@@ -368,39 +368,39 @@ def test_mailing_unique_name(self):
...
             'subject': 'First subject',
             'source_id': self.env['utm.source'].create({'name': 'Custom Source'}).id,
         }, {
             'subject': 'First subject',
-            'name': 'Mailing',
+            'name': 'Mailing1',
         }, {
             'subject': 'Second subject',
-            'name': 'Mailing',
+            'name': 'Mailing2',
```
Comment (reth-odoo) [inline addons/mass_mailing/tests/test_mailing_internals.py:385]: Overall I don't think it makes sense to change this test. Either we do want the mailings to have different names for some reason (I don't think so, after this change).

If the intent of the change is "you should ensure the name is unique yourself", then we should remove duplicates of the same day from the test, as it's now an invalid input.

File: addons/mass_mailing/tests/test_mailing_internals.py [hunk 13]
```diff
@@ -439,6 +439,7 @@ def test_mailing_cron_trigger(self):
             for (test, truth) in [(False, now), (then, then)]:
                 with self.subTest(schedule_date=test):
                     with self.capture_triggers(cron_id) as capt:
+                        common_mailing_values['name'] = f'Knock knock {test}'
```
Comment (reth-odoo) [inline addons/mass_mailing/tests/test_mailing_internals.py:442]: Well then it's not really a common value is it? :D

I'd just put this in the create after `common values`, and remove "Knock knock" from `common values`

File: addons/test_mass_mailing/tests/test_mailing_sms.py [hunk 14]
```diff
@@ -11,22 +11,22 @@
 from odoo.tools import mute_logger
 
 
-@tagged('mass_mailing', 'mass_mailing_sms')
+@tagged('mass_mailing', 'mass_mailing_sms', 'dato_test')
```
Comment (reth-odoo) [inline addons/test_mass_mailing/tests/test_mailing_sms.py:14]: Seems forgotten.

Fyi you can use this syntax to run single modules/classes/tests

`--test-tags /module:class.test` with every step optional
so 
`--test-tags :TestMassSMSInternals` here for example. Or `--test-tags .test_mass_sms_domain`

File: addons/utm/models/utm_medium.py [hunk 15]
```diff
@@ -20,10 +20,11 @@ class UtmMedium(models.Model):
...
     def create(self, vals_list):
-        new_names = self.env['utm.mixin']._get_unique_names(self._name, [vals.get('name') for vals in vals_list])
-        for vals, new_name in zip(vals_list, new_names):
-            vals['name'] = new_name
-        return super().create(vals_list)
+        new_names = []
+        for item in vals_list:
+            if item.get('name'):
+                new_names.append({'name': item['name'].lower()})
+        return super().create(new_names)
```
Comment (reth-odoo) [inline addons/utm/models/utm_medium.py:27]: Wouldn't this mean we lose all values except for the name?
We should modify `vals_list` or create new values from it.

So maybe
```
for vals in vals_list:
    if name := vals.get('name')
        vals['name'] = name.lower()
```

File: addons/utm/models/utm_mixin.py [hunk 16]
```diff
@@ -58,10 +59,10 @@ def tracking_fields(self):
             ('utm_medium', 'medium_id', 'odoo_utm_medium'),
         ]
 
-    def _find_or_create_record(self, model_name, name):
+    def _find_or_create_record(self, model_name, name, lowercase=False):
         """Based on the model name and on the name of the record, retrieve the corresponding record or create it."""
         Model = self.env[model_name]
-
+        name = name.lower() if lowercase else name
```
Comment (reth-odoo) [inline addons/utm/models/utm_mixin.py:65]: This won't work for existing records that weren't lowered right?
Shouldn't you lower in the caller instead?

File: addons/utm/models/utm_mixin.py [hunk 17]
```diff
@@ -38,7 +38,8 @@ def default_get(self, fields):
                     value = request.httprequest.cookies.get(cookie_name)
                 # if we receive a string for a many2one, we search/create the id
                 if field.type == 'many2one' and isinstance(value, str) and value:
-                    record = self._find_or_create_record(field.comodel_name, value)
+                    lowercase = url_param != 'utm_campaign'
```
Comment (reth-odoo) [inline addons/utm/models/utm_mixin.py:41]: Might make sense to set "lower" as a data point in `tracking_fields` instead, otherwise it looks a bit arbitrary (well it is, but it feels weird to do arbitrary stuff in logic parts of the code)

File: addons/utm/models/utm_source.py [hunk 18]
```diff
@@ -17,10 +17,11 @@ class UtmSource(models.Model):
...
     def create(self, vals_list):
-        new_names = self.env['utm.mixin']._get_unique_names(self._name, [vals.get('name') for vals in vals_list])
-        for vals, new_name in zip(vals_list, new_names):
-            vals['name'] = new_name
-        return super().create(vals_list)
+        new_names = []
+        for item in vals_list:
+            if item.get('name'):
+                new_names.append({'name': item['name'].lower()})
+        return super().create(new_names)
```
Comment (reth-odoo) [inline addons/utm/models/utm_source.py:24]: Same as above, we should preserve the other values.

File: addons/utm/tests/test_utm.py [hunk 19]
```diff
@@ -13,13 +16,18 @@ def test_find_or_create_record(self):
         }])
```
Comment (reth-odoo) [inline addons/utm/tests/test_utm.py:16]: Would add a test to demonstrate the actual value of the same, e.g. `self.assertEqual(source_1.name, 'source 1')` and maybe add a comment saying they should always be created in lowercase so when we come back to it it's clear it was intentional.

File: addons/utm/tests/test_utm.py [hunk 20]
```diff
@@ -29,46 +37,76 @@ def test_find_or_create_with_archived_record(self):
...
-            utm_0 = self.env[utm_model].create({'name': 'UTM dup'})
-
-            utm_1, utm_2, utm_3, utm_4, utm_5 = self.env[utm_model].create([{
+        for utm_model in ('utm.source', 'utm.medium'):
+            utm_1, utm_2, utm_3 = self.env[utm_model].create([{
                 'name': 'UTM 1',
             }, {
                 'name': 'UTM 2',
             }, {
                 'name': 'UTM dup',
```
Comment (reth-odoo) [inline addons/utm/tests/test_utm.py:51]: Feels like these two aren't really useful anymore.

**Review by reth-odoo — COMMENTED** [review]
Seems mostly good to me, nice work

Just a few comments :)

File: addons/website_links/tests/test_ui.py (hunk 11 above)
Comment (datonefaridze) [reply addons/website_links/tests/test_ui.py:24]: @reth-odoo i will answer and read review later, but do you know the answer for this: why do we create search here and then later in a js file we invoke search engine. I think js assumes that demo data is loaded, (otherwise how else is it going to retrieve search engine)

Comment (reth-odoo) [reply addons/website_links/tests/test_ui.py:24]: @datonefaridze indeed, I guess this was "fixed" to acomodate demo data as `like` would match any source with "search" in it, I guess with the lowest id first.

Arguably, this is a bug and we should match exactly the name, not anything that contains the name.

Good catch :)
Maybe check with Luc if we want to change that, and if we want to change it in stable

Comment (reth-odoo) [reply addons/website_links/tests/test_ui.py:24]: @datonefaridze nevermind the previous comment, am dumb :monkey: 

Indeed we actively select "Search engine" in the js test, instead of "Search" for some reason. This test is weird either way though :raised_eyebrow: 

Would be good to fix it to select "Search" indeed (or I guess "search" now)

File: addons/utm/models/utm_mixin.py (hunk 16 above)
Comment (datonefaridze) [reply addons/utm/models/utm_mixin.py:65]: but ilike would also find uppercase records, like: FACEBOOK. and it might attach to that. i don't know, i will do it if u insist

**Comment by awa-odoo** [comment]
Replaced by https://github.com/odoo/odoo/pull/172407
